SubmittingPatches-backports.rst

```diff
@@ -0,0 +1,386 @@
+the following two sections: `General principles`_ and `Cherry-picking rules`_.
+
+
+.. contents::
+   :depth: 3
+
+
+General principles
+------------------
+
+To help the people who will review your backport, please state either in the
+backport PR, or in the backport tracker issue, or in the master tracker issue:
+
+1. what bug is fixed
+2. why this fix is the minimal way to do it
+3. why does this need to be fixed in <release>
+
+The above should be followed especially in cases when the backport could be seen
+as introducing, into a stable branch, code that is not related to a particular
+bug or issue.
+
+Rationale: every modification of a stable branch carries a certain risk of
+introducing a regression. To minimize this risk, backports should be as
+straightforward and narrowly-targeted as possible. As a stable release series
+ages, the importance of following these general principles rises.
+
+
+Cherry-picking rules
+--------------------
+
+The following rules, which have been codified from "best practices" developed
+over years of backporting, apply to the actual backport implementation:
+
+* all fixes should land in master first
+* commits to stable branches should be cherry-picked from master
+* before starting to cherry-pick a set of commits from master, grep the master git history for the SHA1 of each master commit (using ``git log --grep``) to check for follow-up fixes. Include any follow-up fixes found in the set of commits to be cherry-picked.
+* cherry-picks must be done using ``git cherry-pick -x``
+* if a commit could not be cherry-picked from master, the commit message must explain why that was not possible
+* the commit message generated by ``git cherry-pick -x`` must not be modified, except to add a "Conflicts" section below the 
+* the "Conflicts" section must mention all files where changes had to made manually (not just conflicts flagged by git)
+* the "Conflicts" section should also describe the manual changes that were made
+* if a change is to be backported to multiple stable branches, a tracker issue is needed
+
+For more information on conflict resolution and writing the "Conflicts" section,
+see `Conflict resolution`_.
+
+For more information on tracker issues, see `Tracker workflow`_.
+
+Adhering to these rules will make your backport easier for reviewers to
+understand. Not adhering to these rules creates additional work for reviewers
+and may cause your backport PR to be rejected.
+
+Notes on the cherry-picking rules
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+What does "all fixes should land in master first" mean? What if I just need to
+fix the issue in <release>?
+
+Merging a fix directly to a stable (release) branch potentially creates
+a regression in the more recent branches (i.e. all later releases and
+``master``), because it's quite possible the issue will still be present in the
+more recent codestreams after such a fix gets merged to an older stable branch.
```

Comment:
Not sure I understand this paragraph. If the fix only applies to an older branch, how could the issue still be present in the newer ones?

Reply:
I agree, it's not clear enough. How about this:

What does "all fixes should land in master first" mean? What if I just need to
fix the issue in <release>?

As the person fixing the issue, you are required to first check whether the issue exists in master. If it does, then the proper course of action is to create a master tracker (see `Tracker workflow`_) and fix the issue in master, first, and only then cherry-pick the fix to the stable branches that have the issue.

If the issue only exists in the stable branch, and not in master, there are several possibilities:

1. it's a regression introduced into the stable branch by a bad backport
2. the issue was fixed in master by some massive refactoring that cannot be backported
3. the issue was already fixed in master by a cherry-pickable commit

In cases 1 and 2, it's permissible to fix the issue directly in the stable branch, subject to the rule "if a commit could not be cherry-picked from master, the commit message must explain why that was not possible".

In case 3, the issue should be handled like any other backport.

Reply:
That sounds much better, thanks!